pkg/querier/queryrange/roundtrip.go

```diff
@@ -255,6 +261,33 @@ func validateMaxEntriesLimits(req *http.Request, reqLimit uint32, limits Limits)
 	return nil
 }
 
+func validateMatchers(req *http.Request, limits Limits, matchers []*labels.Matcher) error {
+	tenants, err := tenant.TenantIDs(req.Context())
+	if err != nil {
+		return httpgrpc.Errorf(http.StatusBadRequest, err.Error())
+	}
+
+	actual := make(map[string]struct{}, len(matchers))
+	for _, m := range matchers {
+		actual[m.Name] = struct{}{}
+	}
+
+	for _, tenant := range tenants {
+		required := limits.RequiredLabelMatchers(req.Context(), tenant)
+		var missing []string
+		for _, label := range required {
+			if _, found := actual[label]; !found {
+				missing = append(missing, label)
+			}
+		}
+
+		if len(missing) > 0 {
+			return fmt.Errorf(logqlmodel.ErrRequiredMatchersMissing, strings.Join(missing, ", "))
+		}
+	}
```

Comment:
I think it would be useful to the end user if in this error message we included the list of labels that were present along with those that are missing? for example in the case of  your binop query in the tests `count_over_time({app="foo"}[1m]) / count_over_time({pod="bar"}[1m] offset 1m)` 

I can imagine it could be difficult to see that one particular side of the query is missing the label when the query is a lot larger (in terms of # of characters)

Reply:
+1